### PR TITLE
Annotate `kube-apiserver` deployment after labeling the resources for encryption

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area control-plane security
/kind bug

**What this PR does / why we need it**:
When the encryption configuration is modified, we label the resources in parallel so that they are rewritten in the ETCD, then snapshot ETCD and remove the label. But in case there are thousands of resources and the client is throttled during the removal of label step, in the next try these resources are again labeled and this results in a cycle.
This PR adds an annotation to the kube-apiserver deployment marking the completion of adding of labels so that we don't repeat this again in case the labels are partially removed from resources.

The task functions for a particular GVK is executed in parallel but sets of tasks for different GVKs are now run in sequence to prevent hitting the rate limit.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Introduced with https://github.com/gardener/gardener/pull/8842
/cc @rfranzke @timuthy as reviewers of the original PR

**Release note**:

```bugfix user
The `kube-apiserver` deployment is annotated to mark the completion of labeling the resources for encrytion so that this step is not repeated in case the "label removal" step fails and resources are partially without the label.
```

### CODE REVIEW

File: pkg/utils/gardener/secretsrotation/constants.go [hunk 1]
```diff
@@ -18,6 +18,9 @@ const (
 	// AnnotationKeyNewEncryptionKeyPopulated is an annotation indicating that the new ETCD encryption key was populated
 	AnnotationKeyNewEncryptionKeyPopulated = "credentials.gardener.cloud/new-encryption-key-populated"
 
+	// AnnotationKeyResourcesLabeled is an annotation indicating the completion of labeling the resources with the credentials.gardener.cloud/key-name label
+	AnnotationKeyResourcesLabeled = "credentials.gardener.cloud/resources-labeled"
+
```
Comment (shafeeqes) [inline pkg/utils/gardener/secretsrotation/constants.go]:
```suggestion
```

**Comment by shafeeqes** [comment]
On a second thought, we could also have used the same `etcd-snapshotted` annotation as well, but it wouldn't make much difference. If snapshotting fails and the flow is retried, we will try to list the resources without the label again (which would be zero), this can be avoided with the current approach.

**Review by timuthy — COMMENTED** [review]
/lgtm

**Comment by timuthy** [comment]
> On a second thought, we could also have used the same etcd-snapshotted annotation as well, it will decrease a patch call for the KAPI deployment. If snapshotting fails and the flow is retried, we will try to list the resources without the label again (which would be zero), this can be avoided with the current approach.

This doesn't sound like a major impact/improvement. It is explicit as it is implemented right now which I prefer.

**Comment by rfranzke** [comment]
I'm not sure whether I'm missing somehting, but wouldn't it make more sense/be even better/more robust to have an annotation per GVK? With the current approach, we will only not try to relabel again when the labelling process was successful for all GVKs, right? However, when there are really many resources, this might not even work in a single step/reconciliation run? WDYT?

**Comment by shafeeqes** [comment]
> With the current approach, we will only not try to relabel again when the labelling process was successful for all GVKs, right?

While listing the resources during retry we use the requirement that the label shouldn't exist,
https://github.com/gardener/gardener/blob/d35f0fece5da8ecd879788433f0e65a2498ae763/pkg/utils/gardener/secretsrotation/etcd.go#L82-L94

**Comment by rfranzke** [comment]
Okay, then let's proceed with this for now. If we see potential for further optimizations, we can also ship them later when we have more clarity or see the need of even doing so. Thank you!

/approve

**Pull request merged**